sycl/include/CL/sycl/buffer.hpp

```diff
@@ -126,6 +139,26 @@ class buffer {
         detail::getNextPowerOfTwo(sizeof(T)), propList);
   }
 
+  // This constructor is a prototype for a future SYCL specification
+  template <class Container, int N = dimensions,
+            typename = EnableIfOneDimension<N>,
+            typename = EnableIfContiguous<Container>>
+  buffer(Container &container, AllocatorT allocator,
```

Comment:
Whether data is written back for iterators is unclear. The spec currently has the contradictory statements "Data is written back to the same iterator set if the iterator is not a const iterator." and "A buffer can be constructed from a pair of iterator values. In this case, [the] destructor will not copy back any data". See https://gitlab.khronos.org/sycl/Specification/issues/162 (restricted to Khronos members).

Independent of what we decide for iterators, I think we definitely want to have copy-back for non-const containers analogous to how it works for the `T* hostData` constructor. And also allow const containers to be passed to the constructor, which then disables write-back. Providing both doesn't require an extra overload because the type deduction automatically deduces the const and non-const versions (this doesn't work for r-values but that isn't a concern here). Thus I don't see a reason to add an extra overload.

I agree we want to support all ranges in the future. In fact I think we should allow already support more than contiguous containers (see other comment). I don't think we want to enable/disable write-back based on whether it is `range` or `view`. `std::ranges::range` doesn't require that it is writable and `std::ranges::view` doesn't apply it isn't. Whether write-back is enabled should depend on whether it is an [`OutputRange`](https://en.cppreference.com/w/cpp/experimental/ranges/range/OutputRange). Note this is only in TS in hasn't been merged yet but I expect something like it will be added.